homeassistant/components/tuya.py

```diff
@@ -0,0 +1,166 @@
+"""
+Support for Tuya Smart devices.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tuya/
+"""
+from datetime import timedelta
+import logging
+import voluptuous as vol
+
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import (CONF_USERNAME, CONF_PASSWORD)
+from homeassistant.helpers import discovery
+from homeassistant.helpers.dispatcher import dispatcher_send
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+REQUIREMENTS = ['tuyapy==0.1.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_COUNTRYCODE = 'country_code'
+
+DOMAIN = 'tuya'
+DATA_TUYA = 'data_tuya'
+
+SIGNAL_DELETE_ENTITY = 'tuya_delete'
+SIGNAL_UPDATE_ENTITY = 'tuya_update'
+
+SERVICE_FORCE_UPDATE = 'force_update'
+SERVICE_PULL_DEVICES = 'pull_devices'
+
+TUYA_TYPE_TO_HA = {
+    'switch': 'switch'
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_COUNTRYCODE): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up Tuya Component."""
+    from tuyapy import TuyaApi
+
+    tuya = TuyaApi()
+    username = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    country_code = config[DOMAIN][CONF_COUNTRYCODE]
+
+    hass.data[DATA_TUYA] = tuya
+    tuya.init(username, password, country_code)
+    hass.data[DOMAIN] = {
+        'entities': {}
+    }
+
+    device_list = tuya.get_all_devices()
+    device_type_list = {}
+    for device in device_list:
+        dev_type = device.device_type()
+        if (dev_type in TUYA_TYPE_TO_HA and
+                device.object_id() not in hass.data[DOMAIN]['entities']):
+            ha_type = TUYA_TYPE_TO_HA.get(dev_type)
+            if ha_type not in device_type_list:
+                device_type_list[ha_type] = []
+            device_type_list[ha_type].append(device.object_id())
+    for ha_type in device_type_list:
+        discovery.load_platform(
+            hass, ha_type, DOMAIN,
+            {'dev_ids': device_type_list.get(ha_type)}, config)
+
+    def poll_devices_update(event_time):
+        """Check if accesstoken is expired and pull device list from server."""
+        _LOGGER.info("Pull Devices From Tuya")
+        devices = tuya.poll_devices_update()
+        if devices is None:
+            return
+        # Add new discover device.
+        device_list = tuya.get_all_devices()
```

Comment:
This block of code until the loading of platforms looks the same as the block above. Extract it into a helper function.